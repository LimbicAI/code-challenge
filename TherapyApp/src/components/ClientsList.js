import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  FlatList,
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
} from 'react-native';

const ClientsList = ({clients, itemOnPress, fetchMore}) => (
  <View style={styles.list}>
    <FlatList
      data={clients}
      keyExtractor={({id}) => `client-${id}`}
      ListEmptyComponent={<Text>Clients list is empty</Text>}
      renderItem={({item: {id, firstName, lastName}}) => (
        <TouchableOpacity
          key={id}
          onPress={() => itemOnPress(id)}
          style={styles.listItem}>
          <Text>
            {firstName} {lastName}
          </Text>
        </TouchableOpacity>
      )}
    />
    <Button title="Fetch more" onPress={fetchMore} />
  </View>
);

const styles = StyleSheet.create({
  list: {
    borderTopWidth: 1,
    borderColor: 'lightgray',
  },
  listItem: {
    padding: 5,
    borderBottomWidth: 1,
    borderColor: 'lightgray',
  },
});

ClientsList.propTypes = {
  clients: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
    }),
  ),
  itemOnPress: PropTypes.func.isRequired,
  fetchMore: PropTypes.func.isRequired,
};

export default ClientsList;
