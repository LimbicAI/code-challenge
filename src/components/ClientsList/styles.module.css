.answerText {
	display: flex;
	flex-flow: row;
}

.answerChoices {
	display: flex;
	flex-flow: column;
}

.client {
	border: 1px solid gray;
	border-radius: 5px;
	cursor: pointer;
	margin: 4px 0;
	padding: 4px 0;
	width: 90%;
	background-color: transparent;
}

.client:hover {
	border-color: #9ad1d4;
	background-color: #9ad1d4;
}

.clientQuestions {
	/* transform: scaleY(0); */
	/* height: 0; */
	opacity: 0;
	flex-flow: column;
	width: 90%;
}

.clientQuestions > div:not(:last-child) {
	border-bottom: 1px solid gray;
}

.expanded {
	background-color: #003249;
	color: white;
	font-weight: 600;
}

.showQuestions {
	display: flex;
	height: auto;
	opacity: 1;
	transform: scaleY(1);
	transition: opacity 1.25s ease-in-out;
}

.label {
	font-weight: 500;
	padding-right: 4px;
}

.list {
	text-align: left;
	margin: 0;
}

.modal {
	display: flex;
	flex-flow: column;
	justify-content: center;
	align-items: center;
	background-color: rgba(0, 0, 0, 0.4);
	position: absolute;
	top: 0;
	left: 0;
	width: 100vw;
	height: 100vh;
}

.question {
	display: flex;
	flex-flow: row;
	justify-content: space-between;
	font-size: 18px;
	font-weight: 500;
	align-items: center;
	width: 80%;
}

.questions {
	display: flex;
	flex-flow: column;
	align-items: center;
}

.questionsAnswers {
	display: flex;
	flex-flow: column;
	align-items: center;
}

.questionsAnswers > div {
	text-align: left;
	width: 100%;
}

@media only screen and (min-width: 600px) {
	.clientQuestions {
		width: 80%;
	}
}

@media only screen and (min-width: 960px) {
	.client {
		width: 60%;
	}

	.clientQuestions {
		width: 60%;
	}
}
