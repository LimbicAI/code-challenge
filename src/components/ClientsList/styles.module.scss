@import '../../styles/colors.scss';

.answerText {
  display: flex;
  flex-flow: row;
}

.answerChoices {
  display: flex;
  flex-flow: column;
}

.client {
  border: 1px solid gray;
  border-radius: 5px;
  cursor: pointer;
  margin: 4px 0;
  padding: 4px 0;
  width: 90%;
  background-color: transparent;
}

.client:hover {
  border-color: $shade2;
  background-color: $shade2;
}

.clientQuestions {
  flex-flow: column;
  width: 90%;
}

.clientQuestions > div:not(:last-child) {
  border-bottom: 1px solid gray;
}

.expanded {
  background-color: $shade4;
  color: white;
  font-weight: 600;
}

.showQuestions {
  animation-name: expand;
  animation-duration: 0.2s;
  display: flex;
  height: auto;
}

.label {
  font-weight: 500;
  padding-right: 4px;
}

.list {
  text-align: left;
  margin: 0;
}

.modal {
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.4);
  position: absolute;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
}

.question {
  display: flex;
  flex-flow: row;
  justify-content: space-between;
  font-size: 18px;
  font-weight: 500;
  align-items: center;
  width: 80%;
}

.questions {
  align-items: center;
  animation-name: fadeIn;
  animation-duration: 0.5s;
  display: flex;
  flex-flow: column;
  margin-top: 16px;
}

.questionsAnswers {
  display: flex;
  flex-flow: column;
  align-items: center;
  padding: 0 8px;
}

.questionsAnswers > div {
  text-align: left;
  width: 100%;
}

@keyframes expand {
  from {
    transform: translateY(50%);
    opacity: 0;
  }

  to {
    transform: translateY(0);
    opacity: 1;
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@media only screen and (min-width: 600px) {
  .clientQuestions {
    width: 80%;
  }
}

@media only screen and (min-width: 960px) {
  .client {
    width: 60%;
  }

  .clientQuestions {
    width: 60%;
  }
}
