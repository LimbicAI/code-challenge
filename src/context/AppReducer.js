const updateLocalStorage = (type, questions) => {
	localStorage.setItem(type, JSON.stringify(questions));
};

const reducer = (state, action) => {
	let updatedQuestions;
	let updatedClients = [];
	switch (action.type) {
		case 'REMOVE_QUESTION':
			updatedQuestions = state.questions.filter((question, index) => {
				return index !== action.payload.index;
			});

			updatedClients = [];
			state.clients.forEach(client => {
				updatedClients.push({
					name: client.name,
					questions: client.questions.filter(
						question => question.q !== action.payload.description
					)
				});
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions,
				clients: updatedClients
			};
		case 'ADD_QUESTION':
			const { description, choices, type } = action.payload;
			const newQuestion = {
				description,
				choices,
				type
			};
			updatedQuestions = [newQuestion, ...state.questions];
			updatedClients = state.clients.map(client => client);
			updatedClients.forEach(client => {
				client.questions.push({ q: description, a: '-' });
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions
			};
		case 'EDIT_QUESTION':
			const updatedQuestion = {
				id: action.payload.id,
				description: action.payload.description,
				type: action.payload.type,
				choices: action.payload.choices
			};

			updatedQuestions = state.questions.map((question, index) => {
				if (index === updatedQuestion.id) {
					return updatedQuestion;
				}
				return question;
			});

			updatedClients = [];
			state.clients.forEach(client => {
				updatedClients.push({
					name: client.name,
					questions: client.questions.map((question, index) => {
						if (action.payload.oldDescription === question.q) {
							return {
								...question,
								q: action.payload.description
							};
						}
						return question;
					})
				});
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions,
				clients: updatedClients
			};

		case 'UPDATE_ANSWERS':
			const clientName = action.payload.client.name;
			const updatedClient = {
				name: clientName
			};
			updatedClient.questions = Object.values(action.payload.answers).map(
				(answer, index) => {
					return {
						q: action.payload.client.questions[index].q,
						a: answer
					};
				}
			);

			const position = state.clients.findIndex(
				client => client.name === clientName
			);
			updatedClients = [...state.clients];
			updatedClients[position] = updatedClient;
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				clients: updatedClients
			};

		default:
			return state;
	}
};

export default reducer;
