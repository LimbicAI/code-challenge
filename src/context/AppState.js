import React, { createContext, useReducer } from 'react';
import AppReducer from './AppReducer';
import CLIENTS from '../fixtures/clients';
import QUESTIONS from '../fixtures/questions';

const savedQuestions = localStorage.getItem('questions');
const savedClients = localStorage.getItem('clients');

const initialState = {
	questions: savedQuestions ? JSON.parse(savedQuestions) : QUESTIONS,
	clients: savedClients ? JSON.parse(savedClients) : CLIENTS
};

export const AppContext = createContext(initialState);

export const AppProvider = ({ children }) => {
	const [state, dispatch] = useReducer(AppReducer, initialState);

	const removeQuestion = payload => {
		dispatch({
			type: 'REMOVE_QUESTION',
			payload
		});
	};

	const addQuestion = question => {
		dispatch({
			type: 'ADD_QUESTION',
			payload: question
		});
	};

	const editQuestion = question => {
		dispatch({
			type: 'EDIT_QUESTION',
			payload: question
		});
	};

	const updateAnswers = (client, answers, questions) => {
		dispatch({
			type: 'UPDATE_ANSWERS',
			payload: { client, answers, questions }
		});
	};

	return (
		<AppContext.Provider
			value={{
				clients: state.clients,
				questions: state.questions,
				removeQuestion,
				addQuestion,
				editQuestion,
				updateAnswers
			}}
		>
			{children}
		</AppContext.Provider>
	);
};
