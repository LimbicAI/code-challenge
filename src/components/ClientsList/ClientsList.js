import React, { useContext, useState } from 'react';
import clsx from 'clsx';
import { AppContext } from '../../context/AppState';
import styles from './styles.module.css';

const QuestionsList = () => {
	const { clients } = useContext(AppContext);
	const [showQuestions, setShowQuestions] = useState(
		clients.map(() => false)
	);

	return (
		<div className={styles.questions}>
			{clients.map((client, index) => (
				<React.Fragment key={`${client.name}_${index}`}>
					<button
						className={styles.client}
						onClick={() => {
							const updatedShowQuestions = clients.map(
								() => false
							);
							updatedShowQuestions[index] = !showQuestions[index];
							setShowQuestions(updatedShowQuestions);
						}}
					>
						{client.name}
					</button>
					<div
						className={clsx(styles.clientQuestions, {
							[styles.showQuestions]: showQuestions[index]
						})}
					>
						{client.questions.map(question => (
							<div
								className={styles.questionsAnswers}
								key={client.name + question.q}
							>
								<p>
									<span className={styles.label}>Q: </span>
									{question.q}
								</p>
								<p>
									<span className={styles.label}>A: </span>
									{question.a}
								</p>
							</div>
						))}
					</div>
				</React.Fragment>
			))}
		</div>
	);
};

export default QuestionsList;
