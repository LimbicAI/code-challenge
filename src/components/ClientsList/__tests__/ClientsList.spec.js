import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { AppProvider } from '../../../context/AppState';
import ClientsList from '../ClientsList';
import CLIENTS from '../../../fixtures/clients';

let result;

describe('test', () => {
	beforeEach(() => {
		result = render(
			<AppProvider>
				<ClientsList />
			</AppProvider>
		);
	});
	it('matches the snapshot', () => {
		expect(result.asFragment()).toMatchSnapshot();
	});

	it('expands the question list on name click', () => {
		expect(
			screen.queryByText(CLIENTS[0].questions[0].q)
		).not.toBeInTheDocument();
		fireEvent.click(screen.getByText(CLIENTS[0].name));
		expect(
			screen.queryByText(CLIENTS[0].questions[0].q)
		).toBeInTheDocument();
	});
});
