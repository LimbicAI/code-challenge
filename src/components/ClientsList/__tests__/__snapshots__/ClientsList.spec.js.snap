// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="questions"
  >
    <button
      class="client"
    >
      Andy
    </button>
    <button
      class="client"
    >
      James
    </button>
    <button
      class="client"
    >
      Rick
    </button>
  </div>
</DocumentFragment>
`;
